### PR TITLE
[REVIEW] Parquet reader: fix empty string potentially read as null

### LINKED ISSUE #4286: [BUG] Parquet reader can load non-null strings as null values
**Describe the bug**
Reading a Parquet file with strings sometimes is returning null entries for strings that are not null.  Spark loads the file and shows there are no nulls, but libcudf loads the file containing one or more nulls.  It appears the values that can become null are supposed to be loaded as empty strings.

**Steps/Code to reproduce bug**
I am not sure yet how to forcefully trigger it with a unit test, but I have some Parquet files that demonstrate the issue.  I can provide them upon request.

**Expected behavior**
libcudf Parquet reader should produce the same data as Spark's Parquet reader.

### PR DESCRIPTION
**Pull request by OlivierNV**
Data offset may be incorrectly considered out-of-range
Fixes #4286

### CODE REVIEW

**Comment by jlowe** [comment]
I haven't had a chance to dig into the code details.  However I did a quick test with this patch and verified that it does appear to solve the problem.

**Comment by OlivierNV** [comment]
> I haven't had a chance to dig into the code details. However I did a quick test with this patch and verified that it does appear to solve the problem.

Good, because I dug in the code but did not verify it actually solved the problem :)

**Comment by OlivierNV** [comment]
rerun tests

File: cpp/src/io/parquet/page_data.cu [hunk 1]
```diff
@@ -663,7 +663,7 @@ inline __device__ void gpuOutputString(volatile page_state_s *s, int src_pos, vo
     {
         // Plain encoding
         uint32_t dict_pos = s->dict_idx[src_pos & (NZ_BFRSZ - 1)];
-        if (dict_pos < (uint32_t)s->dict_size)
+        if (dict_pos <= (uint32_t)s->dict_size)
```
[the remark this answers is not in the document]
Comment (OlivierNV) [reply cpp/src/io/parquet/page_data.cu]: IIRC, the length is checked before dict_idx is filled-in, so here it just does an additional check to make sure that the starting position is within range to check the most common failure cases. There are unfortunately many ways malformed parquet files can still cause bad reads, for example negative lengths, or values that overflow 32-bit, but difficult at that relatively late stage to report it back to the host.
I think reliably detecting parquet errors and return a status would have to be a separate issue & PR (we'd have to keep track of a per-page error status)

[the remark this answers is not in the document]
Comment (jlowe) [reply cpp/src/io/parquet/page_data.cu]: > I think reliably detecting parquet errors and return a status would have to be a separate issue

Totally agree.  The fix just made me think about malformed data and whether it was being handled today or not.

**Pull request merged**